stdlib/json.gr

```diff
@@ -0,0 +1,2308 @@
+    },
+  }: JSONWriter
+}
+
+/**
+ * Prints the JSON object into a string with specific formatting settings.
+ * 
+ * @param json: The JSON object to print
+ * @param format: Custom formatting setttings
+ * @returns A `Result` object with either the string containing the printed JSON or an error if the input object cannot be represented in the JSON format.
+ *
+ * @example print(Result.unwrap(toString(JSONObject([("currency", JSONString("€")), ("price", JSONNumber(99.9))]), defaultCompactAndSafeFormat())))
+ * 
+ * Example output:
+ * ```json
+ * {"currency":"\u20ac","price":99.9}
+ * ```
+ */
+export let toString = (json: JSON, format: FormattingSettings) => {
+  let buf = Buffer.make(16)
+
+  let writer = makePrettyPrintJSONWriter(format, buf)
+
+  let error = writer.emit(json)
+
+  match (error) {
+    None => Ok(Buffer.toString(buf)),
+    Some(e) => Err(e),
+  }
+}
+
+/**
+ * Prints the JSON object into a string with compact formatting optimized for
+ * small size. Recommended for all uses where the intended consumer is a
+ * machine program. For example in REST APIs.
+ * 
+ * Using this function can be preferred over `toString` with compact formatting
+ * settings since it can be slightly faster.
+ * 
+ * @param json: The JSON object to print
+ * @returns A `Result` object with either the string containing the printed JSON or an error if the input object cannot be represented in the JSON format.
+ *
+ * @example print(Result.unwrap(toStringCompact(JSONObject([("currency", JSONString("€")), ("price", JSONNumber(99.9))]))))
+ * 
+ * Example output:
+ * ```json
+ * {"currency":"€","price":99.9}
+ * ```
+ */
+export let toStringCompact = (json: JSON) => {
+  let buf = Buffer.make(16)
+
+  let writer = makeCompactJSONWriter(buf)
+
+  let error = writer.emit(json)
+
+  match (error) {
+    None => Ok(Buffer.toString(buf)),
+    Some(e) => Err(e),
+  }
+}
+
+/**
+ * Prints the JSON object into a string with default pretty formatting
+ * settings. Recommended for uses where the intended consumer is a person or
+ * the text should be easily inspectable. For example configuration files or
+ * document file formats.
+ * 
+ * @param json: The JSON object to print
+ * @returns A `Result` object with either the string containing the printed JSON or an error if the input object cannot be represented in the JSON format.
+ * 
+ * Example output:
+ * ```json
+ * {
+ *   "currency": "€",
+ *   "price": 99.9
+ * }
+ * ```
+ *
+ * @example print(Result.unwrap(toStringPretty(JSONObject([("currency", JSONString("€")), ("price", JSONNumber(99.9))]))))
+ */
+export let toStringPretty = (json: JSON) =>
+  toString(json, defaultPrettyFormat())
+
+/**
+ * Represents errors for JSON parsing along with a human readable text message.
+ */
+export enum JSONParseError {
+  UnexpectedEndOfInput(String),
+  UnexpectedToken(String),
+  InvalidUTF16SurrogatePair(String),
+}
+
+/**
+ * Internal data structure used during parsing.
+ */
+record JSONParserImplHelper {
```

Comment:
```suggestion
record JSONParserState {
```
(I also recommend changing the `parserImplHelper` arg names below)

Reply:
Renamed.